### ISSUE TITLE
Create LSP for Fish

### ISSUE THREAD

**Issue by EmilySeville7cfg** [bmalehorn/vscode-fish#36]
From LSP I expect:

- syntax and semantic validation (highlight not available variables and functions in the current scope, tell about non-existing flags for commands)
- variable and function suggestions available in the current scope
- ability to rename variables and functions

Most of these features exist for Bash LSP and Nushell LSP.

**Comment by zanchey** [comment]
Some work done in https://github.com/ndonfris/fish-lsp

**Comment by ndonfris** [comment]
Ya @bmalehorn spot on with the summary. 👍 

Apologies for not being active here yet, but I agree that making the fish-lsp compatible with this extension makes sense. Most of the features requested in @EmilyGraceSeville7cf's initial issue request are supported in the `fish-lsp` and have been for quite some time. 

Recently, while I've been [trying to improve the syntax highlighting](https://github.com/ndonfris/fish-lsp/pull/113) support in the [VSCode extension](https://github.com/ndonfris/vscode-fish-lsp) of the server recently, it has become apparent to me supporting an in-depth [semantic-token handler](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokensParams) is not a good fit for the language server itself, especially given the nature of a language server switching out language clients that likely already have their own custom ecosystem independent method of providing highlights source code.

A common approach I've come across is to use a mature language client's existing syntax highlighting API, and then layering the tokens that analysis is required for on top of the existing syntax highlighting.

For most common [definition symbols](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentSymbol) in fish shell, the server already provides easy access to analyze and extract relevant semantic tokens for variable and function definitions.

Some examples:

```fish
set -lx foo "bar"  # 'set' would be a keyword, 'foo' a variable
#       ^^^--- variable definition
for i in (seq 1 10)  # 'for' would be a keyword, 'i' a variable
#   ^--- variable definition
end
argparse h/help -- "$argv"  # 'argparse' would be a keyword
#        ^ ^^^^--- variable definition
export baz='bar' 
#      ^^^--- variable definition
# etc...
```

Now of course, I could parse out other tokens too, but translating [`SyntaxNode[]`](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/src/node.ts) items to [`SemanticToken[]`](https://github.com/microsoft/vscode-languageserver-node/blob/7e14d6277c655680d8603edce4beae818b9e70f0/types/src/main.ts#L4026) can be quite tedious to correctly support things like: nested syntax's, special tokens, token modifier resolution and other edge cases (due to the [languageserver-protocols requirement that `SemanticToken[]`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens) items must be a single array of integers).

Given the popularity of this extension, I am proposing that semantic-token support in the [`fish-lsp`](https://github.com/ndonfris/fish-lsp?tab=readme-ov-file) prioritizes compatibility with the language structure defined in this extension (if they aren't already using tree-sitter). Thus, resulting in the server having a strictly defined semantic token protocol that is tested across many users, without introducing inconsistent theme related issues that would be possible if the server used its own custom semantic token legend.

Assuming this approach is acceptable, we'd really only need to confirm a few things:

- Mainly, which nodes/symbols would be most beneficial to provide as semantic tokens from the server? I imagine variable and function definitions would be a good start, but are there any other nodes that would be useful to provide as semantic tokens?

- Secondly, what [`SemanticTokenModifiers`](https://github.com/microsoft/vscode-languageserver-node/blob/7e14d6277c655680d8603edce4beae818b9e70f0/types/src/main.ts#L3995) are relevant to this extension? For example, I can provide the declaration scope for [`FishSymbol`](https://github.com/ndonfris/fish-lsp/blob/cc77cf3f0cf7946486ae26b64e5a176de5642098/src/parsing/symbol.ts#L754) items, and can use the analyzed fish shell workspace to determine which commands are built-in, auto-loaded, user-defined, etc. When it comes to any other tokens though, I've found that clients might treat token modifiers differently, so it would be good to get some feedback on any edge case modifiers before implementing them. 

Looking forward to further discussion on this topic. Will try to find time to go through the [`fish.tmLanguage.json`](https://github.com/bmalehorn/vscode-fish/blob/master/syntaxes/fish.tmLanguage.json) in the meantime.

**Comment by bmalehorn** [comment]
I looked into `fish-lsp` and got a basic version working. I have a PR over here with my current work for those interested: https://github.com/bmalehorn/vscode-fish/pull/55

Right now, I am most focused on getting `fish-lsp` working at all & releasing it with what it supports today.

### highlighting


@ndonfris thanks for making `fish-lsp` and making this in-depth post about semantic highlighting. I'm definitely more than happy to incorporate `fish-lsp` and I'm flattered by the offer to make more changes for this extension!

I will probably think about highlighting & semantic tokens more deeply once I have `fish-lsp` released as part of `vscode-fish`. But here's my thoughts:

* syntax highlighting: This extension uses textmate language definition, which VSCode has an engine for which parses the file start to finish with a series of regexes and simple state machine. The only purpose is syntax highlighting, it has no concept that `$foo` on line 2 is the same variable set as `foo` on line 1, it just knowns that it matches `\$\w+` so it's a variable.
* IMO the "syntax highlight" is already pretty OK. For things like highlighting strings, keywords, variables, etc., I imagine we'll continue to use the existing syntax highlighting.
* VSCode has some ability to do "syntax highlighting" first, and then "semantic highlighting" on top of that.
* The purpose of an LSP for highlighting would be to add "semantic highlighting", not "syntax highlighting". For that reason I imagine I won't need everything little thing labeled from fish-lsp, but only the things I'll want to add on top of syntax highlighting.
* for instance, in TypeScript, semantic highlighting is responsible for making functions, types, variables, and constants all colored differently. This has to come from semantic highlighting because syntax highlighting cannot look at `import { foo } from 'bar'` and say if `foo` is a function, type, variable or constant. I'm not sure what else it's responsible for.


So:
> Mainly, which nodes/symbols would be most beneficial to provide as semantic tokens from the server? I imagine variable and function definitions would be a good start, but are there any other nodes that would be useful to provide as semantic tokens?

> Secondly, what [SemanticTokenModifiers](https://github.com/microsoft/vscode-languageserver-node/blob/7e14d6277c655680d8603edce4beae818b9e70f0/types/src/main.ts#L3995) are relevant to this extension? For example, I can provide the declaration scope for [FishSymbol](https://github.com/ndonfris/fish-lsp/blob/cc77cf3f0cf7946486ae26b64e5a176de5642098/src/parsing/symbol.ts#L754) items, and can use the analyzed fish shell workspace to determine which commands are built-in, auto-loaded, user-defined, etc. When it comes to any other tokens though, I've found that clients might treat token modifiers differently, so it would be good to get some feedback on any edge case modifiers before implementing them.

Yes, variable & function definitions sound useful. However, I am not really sure what providing this semantic information will do in the editor though. Will functions get highlighted in a different color? Will something happen when I hover over them?

I understand you don't use VSCode on a regular basis. But, what does adding these semantic tokens do in other editors?

Is there language server documentation about why you might want to send this information, and what it does?

Thanks for bearing with me. I'm learning the LSP interface as we go.

**Comment by ndonfris** [comment]
> Right now, I am most focused on getting fish-lsp working at all & releasing it with what it supports today.

Thanks for testing and improving fish-lsp!

I originally planned to have [vscode-fish-lsp](https://github.com/ndonfris/vscode-fish-lsp/) depend on this extension via [`extensionDependencies`](https://code.visualstudio.com/api/references/extension-manifest) - this matches the [recommended pattern](https://code.visualstudio.com/api/language-extensions/overview) for tracking issues per project.

However, I'm fine with encapsulating fish-lsp entirely in this project instead and deprecating my extension. If we go that route, we'd need a GitHub action to auto-bump the fish-lsp dependency and publish new releases when fish-lsp updates.

Either way works - let me know your preference! I'll add more details about the server implementation on #55.

___

> what does adding these semantic tokens do in other editors?

**TL;DR:** Semantic tokens provide editor-agnostic syntax highlighting via LSP. The server can handle all highlighting without requiring TextMate grammars (VSCode) or tree-sitter highlights (Neovim).

This is not necessarily a good idea though, because different editors might have different expectations for how semantic tokens are named & used.

> Yes, variable & function definitions sound useful. However, I am not really sure what providing this semantic information will do in the editor though. Will functions get highlighted in a different color?

Yes, function names <ins>**could potentially**</ins> get highlighted in a different color, and variable names would for the most part get highlighted differently as well.

<div align="center">
<table>
<tr><th>No Semantics</th><th>With Semantic Tokens</th></tr>
<tr><td><div align="left">

```fish
set -l my_var 10
#      ^^^^^^ not highlighted
```

</div></td><td><div align="left">

```fish
set -l my_var 10
#      ^^^^^^ highlighted as variable
```

</div></td></tr>
</table>
</div>

> [!NOTE]
> Current behavior depends on the editor and theme.

**My current goal:**

Allow an opt-in setting for the server to provide different levels of semantic token support:

1. **Default:** Only provide semantic tokens for function and variable definition names
   (with potential modifier support)
2. **Extended (opt-in):** Follow this plugin's naming conventions for consistent cross-editor theming.
   This would allow themes that other editors use to highlight the same way this plugin does, by treating all syntax tokens the way this plugin does.

**Potential issues:**

The LSP spec uses modifiers like `declaration`, `definition`, `readonly`, etc. Fish doesn't distinguish between declaration/definition - everything is just a definition. This can cause unexpected colors like the following:

- Function name: blue
- Variable name: yellow
- Definition modifier: gray

Now both function and variable names could end up gray instead of their expected colors. [Potential workaround?](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#custom-token-types-and-modifiers) Themes might also use the blue/yellow colors for each token type respectively.

Was particularly wondering if you had any recommendations on the best way to define the modifiers to avoid this issue.

<details>
<summary>Example of potential issue from TypeScript in VSCode</summary>

<img width="586" height="311" alt="Image" src="https://github.com/user-attachments/assets/1599b6f5-aa57-4477-9c0a-749f9502c53f" />

Different editors may expect different modifier naming conventions beyond VSCode's defaults (e.g., `storage.modifier`).
</details>

> Will something happen when I hover over them?

Only highlight colors should. If you wanted to see more hover info, you could use the tooltip from the command: `Ctrl+shift+p` + `Developer: Inspect Editor Tokens and Scopes`.

This is the main way I've found to test how the semantic tokens are being applied in vscode, and can be tedious to check during development. Recommendations for easier testing methods would be awesome :)

> Is there language server documentation about why you might want to send this information, and what it does?

Assuming this question is about the LSP specification regarding semantic token modifiers, I interpret the wording of the docs to allow for providing more exact info about a token being used.

Modifiers enable finer distinctions (e.g., builtin functions like `fish_add_path` vs user functions like `my_custom_function`, or `global` vs `local` variables). Ideally, I thought it'd make sense to allow the theme to determine if/how these modifiers should be highlighted, and highlight them as their token type if not specified otherwise. Maybe it doesn't make sense to use modifiers at all for fish - open to suggestions!

The docs [here](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#custom-textmate-scope-mappings) seem particularly relevant.
